File: 2019/py/day_08.py
from pathlib import Path


def main():
    text = (Path(__file__).parent / "../input/input_08.txt").read_text().strip()
    width, height = 25, 6
    lsize = width * height
    layers = [[int(x) for x in text[i:i+lsize]] for i in range(0, len(text), lsize)]
    target = min(layers, key=lambda l: sum(c == 0 for c in l))
    print("Part 1:", sum(x == 1 for x in target) * sum(x == 2 for x in target))
    print("Part 2:")
    for row in range(height):
        for col in range(width):
            for layer in layers:
                if (c := layer[row * width + col]) != 2:
                    print('#' if c else ' ', end='')
                    break
        print()


if __name__ == "__main__":
    main()

File: rust/2019/src/bin/day_08.rs
use std::fs;

const WIDTH: usize = 25;
const HEIGHT: usize = 6;
const LSIZE: usize = WIDTH * HEIGHT;

fn main() {
    let text = fs::read_to_string("../input/input_08.txt").unwrap();
    let data = text.trim().as_bytes();
    let layers: Vec<_> = (0..data.len()).step_by(LSIZE)
        .map(|i| &data[i..i+LSIZE])
        .collect();
    let min_layer = layers.iter().min_by_key(|layer| layer.iter().filter(|&&b| b == b'0').count()).unwrap();
    let prod: usize = (b'1'..=b'2').map(|t| min_layer.iter().filter(|&&b| b == t).count()).product();
    println!("Part 1: {}", prod);
    println!("Part 2:");
    for row in 0..HEIGHT {
        for col in 0..WIDTH {
            for layer in &layers {
                match layer[row * WIDTH + col] {
                    b'0' => print!(" "),
                    b'1' => print!("#"),
                    _ => continue,
                }
                break;
            }
        }
        println!();
    }
}
